### ISSUE TITLE
Contract Review

### ISSUE THREAD

**Issue by docmarionum1** [docmarionum1/bitku#8]
The following is the feedback from the initial review by Rhea Myers at Dapper Labs:

# General

* Contracts, transactions and script should use the named path pattern
- https://docs.onflow.org/cadence/design-patterns/#named-value-field

* Storage paths should be distinctively named, e.g.
`/public/BitkuHaikuCollection` .

# contracts

* Out of interest, do you have to change any gas limits to deploy the contracts?

## HaikuNFT.cdc

* To optimize gas usage, I recommend storing the seed in the NFT, and
making `NFT.text` into a member function of the NFT that returns the
result of calling `generateHaiku()`. This can be run in scripts and
cached. Or optionally cached in a string in the NFT by the owner
calling a function that does so separately from minting - this could
be included in the minting transaction by default for now. If these
approaches do not work for the token's workflows, optimizing the
string functions any further than they already have been will help
save gas.

1..3:

* Make sure to update this comment to reflect the file's name and
status. Comments are visible in deployed contracts on Flow.

30: `        pub var text: String`

* This is publicly settable by the current token owner.

38: `    pub resource interface HaikuCollectionPublic {`

* It's best to implement all the CollectionPublic methods in
interfaces that complement them. It should be `pub let` if it is
immutable, or if the string is meant to be updatable by an admin it
should be `access(self) var` with a `pub` getter on the NFT resource
and an `access(contract)` function to allow admin code to change it.

318: ` init(charityAddress: Address) {`

* Tab should be spaces.

351: ` }`

* Tab should be spaces.

# scripts

## CheckCollection.cdc

* This should take the address as an argument, and can be simplified
to return a boolean rather than a string array, e.g. (untested):
```
pub fun main(address: Address): Bool {
  let account = getAccount(address)

  let capability = account.getCapability(/public/HaikuCollection)

  return capability.borrow<&{HaikuNFT.HaikuCollectionPublic,
NonFungibleToken.CollectionPublic}>() != nil
}
```

# transactions

## transfer_tokens.cdc

* Contract imports should not hardcode emulator addresses.

**Comment by docmarionum1** [comment]
> To optimize gas usage, I recommend storing the seed in the NFT, and
> making NFT.text into a member function of the NFT that returns the
> result of calling generateHaiku(). This can be run in scripts and
> cached. Or optionally cached in a string in the NFT by the owner
> calling a function that does so separately from minting - this could
> be included in the minting transaction by default for now. If these
> approaches do not work for the token's workflows, optimizing the
> string functions any further than they already have been will help
> save gas.

Is there a way to guarantee that `generateHaiku(_ seed: UInt64)` can't be change by me later so that any seed will _always_
return the same text? 

> Out of interest, do you have to change any gas limits to deploy the contracts?

Nope, not to deploy. It uses whatever gas limit is set with `flow project deploy --network=testnet --update`.

> 38: pub resource interface HaikuCollectionPublic {
> It's best to implement all the CollectionPublic methods in
> interfaces that complement them.

What do you mean by this? 

> CheckCollection.cdc, transfer_tokens.cdc

I deleted this as they were leftover from development testing. The place to look for the scripts and transactions as they are being used by the application is `/web/src/cadence/*.js`. I wasn't sure if there was a better way to write scripts and transactions for integration into a frontend using `fcl`. I saw this pattern in some of the examples.

**Comment by docmarionum1** [comment]
> > > 38: pub resource interface HaikuCollectionPublic {
> > > It's best to implement all the CollectionPublic methods in
> > > interfaces that complement them.
> >
> > What do you mean by this?
> 
> Ah my apologies, that was clearer in my head than in the email!
> 
> If you look at the TopShot interface that contains the additional
> public functions for its Collection type:
> 
> https://github.com/dapperlabs/nba-smart-contracts/blob/master/contracts/TopShot.cdc#L525
> 
> You will notice that it also implements all the
> NonFungibleToken.PublicCollection functions:
> 
> https://github.com/onflow/flow-nft/blob/master/contracts/NonFungibleToken.cdc#L104
> 
> I recommend this approach, as it makes it easier to use the Collection
> in some circumstances.

Addressed by: 7e7bf67518d72c0a2b557f1fcc7b90ea74d57151
